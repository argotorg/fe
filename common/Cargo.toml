[package]
name = "fe-common"
version = "0.3.0-alpha"
authors = ["Ethereum Foundation <snakecharmers@ethereum.org>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
tiny-keccak = { version = "2.0", features = ["keccak"] }
hex = "0.4"
codespan-reporting = "0.11.1"
serde = { version = "1", features = ["derive"] }