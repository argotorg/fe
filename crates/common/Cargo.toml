[package]
name = "fe-common"
version = "0.12.0-alpha"
authors = ["The Fe Developers <snakecharmers@ethereum.org>"]
edition = "2021"
license = "Apache-2.0"
repository = "https://github.com/ethereum/fe"

[dependencies]
fe-library = {path = "../library", version = "^0.12.0-alpha"}
tiny-keccak = { version = "2.0", features = ["keccak"] }
hex = "0.4"
codespan-reporting = "0.11.1"
serde = { version = "1", features = ["derive"] }
ron = "0.5.1"
difference = "2.0"
num-traits = "0.2.14"
once_cell = "1.8.0"
include_dir = "0.6.0"
indexmap = "1.6.2"
smol_str = "0.1.21"
