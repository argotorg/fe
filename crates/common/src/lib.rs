pub mod config;
pub mod core;
pub mod diagnostics;
pub mod file;
pub mod indexmap;
pub mod ingot;
pub mod urlext;

use file::FileIndex;

#[salsa::db]
// Each database must implement InputDb explicitly with its own storage mechanism
pub trait InputDb: salsa::Database {
    fn file_index(&self) -> FileIndex;
}

#[doc(hidden)]
pub use paste::paste;

/// Macro for implementing the InputDb trait for a Salsa database struct
///
/// This assumes the database has a field named `index` of type `Option<FileIndex>`.
///
/// # Example
///
/// ```
/// #[derive(Clone)]
/// #[salsa::db]
/// pub struct MyDatabase {
///     storage: salsa::Storage<Self>,
///     index: Option<FileIndex>,
/// }
///
/// impl_input_db!(MyDatabase);
/// ```
#[macro_export]
macro_rules! impl_input_db {
    ($db_type:ty) => {
        #[salsa::db]
        impl $crate::InputDb for $db_type {
            fn file_index(&self) -> $crate::file::FileIndex {
                self.index.clone().expect("FileIndex not initialized")
            }
        }
    };
}

/// Macro for implementing Default for a Salsa database with FileIndex
///
/// This assumes the database has a field named `index` of type `Option<FileIndex>`,
/// and will initialize it properly.
///
/// # Example
///
/// ```
/// #[derive(Clone)]
/// #[salsa::db]
/// pub struct MyDatabase {
///     storage: salsa::Storage<Self>,
///     index: Option<FileIndex>,
/// }
///
/// impl_db_default!(MyDatabase);
/// ```
#[macro_export]
macro_rules! impl_db_default {
    ($db_type:ty) => {
        impl Default for $db_type
        where
            $db_type: $crate::core::HasBuiltinCore,
        {
            fn default() -> Self {
                let mut db = Self {
                    storage: salsa::Storage::default(),
                    index: None,
                };
                let index = $crate::file::FileIndex::default(&db);
                db.index = Some(index);
                $crate::core::HasBuiltinCore::initialize_builtin_core(&mut db);
                db
            }
        }
    };
}

/// Macro for creating a standard Salsa database with FileIndex support
///
/// This creates a struct with:
/// - A `storage` field for Salsa
/// - An `index` field for the FileIndex
/// - Default implementation that initializes the FileIndex
/// - Implementation of InputDb via impl_input_db!
///
/// # Example
///
/// ```
/// define_input_db!(MyDatabase);
/// ```
#[macro_export]
macro_rules! define_input_db {
    ($db_name:ident) => {
        #[derive(Clone)]
        #[salsa::db]
        pub struct $db_name {
            storage: salsa::Storage<Self>,
            index: Option<$crate::file::FileIndex>,
        }

        #[salsa::db]
        impl salsa::Database for $db_name {
            fn salsa_event(&self, _event: &dyn Fn() -> salsa::Event) {}
        }

        $crate::impl_input_db!($db_name);
        $crate::impl_db_default!($db_name);
    };
}

#[macro_export]
macro_rules! impl_db_traits {
    ($db_type:ty, $($trait_name:ident),+ $(,)?) => {
        #[salsa::db]
        impl salsa::Database for $db_type {
            fn salsa_event(&self, _event: &dyn Fn() -> salsa::Event) {}
        }
    };
}
