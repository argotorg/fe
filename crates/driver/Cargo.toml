[package]
name = "fe-driver"
version = "0.13.0-alpha"
authors = ["The Fe Developers <snakecharmers@ethereum.org>"]
edition = "2021"
license = "GPL-3.0-or-later"
repository = "https://github.com/ethereum/fe"

[features]
# Includes the solidity compiler for Yul -> bytecode compilation.
solc-backend = ["fe-yulc"]

[dependencies]
serde_json = "1.0"
fe-abi = {path = "../abi", version = "^0.13.0-alpha"}
fe-analyzer = {path = "../analyzer", version = "^0.13.0-alpha"}
fe-common = {path = "../common", version = "^0.13.0-alpha"}
fe-lowering = {path = "../lowering", version = "^0.13.0-alpha"}
fe-mir = {path = "../mir", version = "^0.13.0-alpha"}
fe-parser = {path = "../parser", version = "^0.13.0-alpha"}
fe-yulgen = {path = "../yulgen", version = "^0.13.0-alpha"}
fe-yulc = {path = "../yulc", version = "^0.13.0-alpha", features = ["solc-backend"], optional = true}
indexmap = "1.6.2"
vfs = "0.5.1"
smol_str = "0.1.21"
toml = "0.5.8"
