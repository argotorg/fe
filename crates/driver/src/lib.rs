use fe_analyzer::context::Analysis;
use fe_analyzer::namespace::items::{IngotId, ModuleId};
use fe_common::diagnostics::Diagnostic;
use fe_common::files::{FileStore, SourceFileId};
use fe_parser::ast::SmolStr;
use fe_yulgen::Db;
use indexmap::IndexMap;
#[cfg(feature = "solc-backend")]
use serde_json::Value;
use std::ops::Deref;

/// The artifacts of a compiled module.
pub struct CompiledModule {
    pub src_ast: String,
    pub lowered_ast: String,
    pub contracts: IndexMap<String, CompiledContract>,
}

/// The artifacts of a compiled contract.
pub struct CompiledContract {
    pub json_abi: String,
    pub yul: String,
    #[cfg(feature = "solc-backend")]
    pub bytecode: String,
}

#[derive(Debug)]
pub struct CompileError(pub Vec<Diagnostic>);

/// Compiles a single input file.
///
/// If `with_bytecode` is set to false, the compiler will skip the final Yul ->
/// Bytecode pass. This is useful when debugging invalid Yul code.
pub fn compile_module(
    files: &FileStore,
    file_id: SourceFileId,
    deps: &IndexMap<SmolStr, Vec<SourceFileId>>,
    _with_bytecode: bool,
    _optimize: bool,
) -> Result<CompiledModule, CompileError> {
    let mut errors = vec![];

    let db = Db::default();

    let Analysis {
        value: module_id,
        diagnostics: parser_diagnostics,
    } = ModuleId::try_new(&db, files, file_id, deps).map_err(CompileError)?;
    errors.extend(parser_diagnostics.deref().clone());

    match fe_analyzer::analyze_module(&db, module_id) {
        Ok(_) => {}
        Err(diagnostics) => {
            errors.extend(diagnostics.into_iter());
            return Err(CompileError(errors));
        }
    };

    if !errors.is_empty() {
        // There was a non-fatal parser error (eg missing parens in a fn def `fn foo: ...`)
        return Err(CompileError(errors));
    }

    // build abi
    let json_abis = fe_abi::build(&db, module_id).expect("failed to generate abi");

    // lower the AST
    let lowered_module_id = fe_lowering::lower_module(&db, module_id);
    let lowered_ast = format!("{:#?}", &lowered_module_id.ast(&db));

    fe_analyzer::analyze_module(&db, lowered_module_id).expect("failed to analyze lowered AST");

    // compile to yul
    let yul_contracts = fe_yulgen::compile(&db, lowered_module_id);

    // compile to bytecode if required
    #[cfg(feature = "solc-backend")]
    let bytecode_contracts = if _with_bytecode {
        match fe_yulc::compile(yul_contracts.clone(), _optimize) {
            Err(error) => {
                for error in serde_json::from_str::<Value>(&error.0)
                    .expect("unable to deserialize json output")["errors"]
                    .as_array()
                    .expect("errors not an array")
                {
                    eprintln!(
                        "Error: {}",
                        error["formattedMessage"]
                            .as_str()
                            .expect("error value not a string")
                            .replace("\\\n", "\n")
                    )
                }
                panic!("Yul compilation failed with the above errors")
            }
            Ok(contracts) => contracts,
        }
    } else {
        IndexMap::new()
    };

    // combine all of the named contract maps
    let contracts = json_abis
        .keys()
        .map(|name| {
            (
                name.to_owned(),
                CompiledContract {
                    json_abi: json_abis[name].to_owned(),
                    yul: yul_contracts[name].to_owned(),
                    #[cfg(feature = "solc-backend")]
                    bytecode: if _with_bytecode {
                        bytecode_contracts[name].to_owned()
                    } else {
                        "".to_string()
                    },
                },
            )
        })
        .collect::<IndexMap<_, _>>();

    Ok(CompiledModule {
        src_ast: format!("{:?}", module_id.ast(&db)),
        lowered_ast,
        contracts,
    })
}

/// Compiles a set of input files.
///
/// If `with_bytecode` is set to false, the compiler will skip the final Yul ->
/// Bytecode pass. This is useful when debugging invalid Yul code.
pub fn compile_ingot(
    name: &str,
    files: &FileStore,
    file_ids: &[SourceFileId],
    deps: &IndexMap<SmolStr, Vec<SourceFileId>>,
    _with_bytecode: bool,
    _optimize: bool,
) -> Result<CompiledModule, CompileError> {
    let mut errors = vec![];

    let db = Db::default();

    let Analysis {
        value: ingot_id,
        diagnostics: parser_diagnostics,
    } = IngotId::try_new(&db, files, name, file_ids, deps).map_err(CompileError)?;
    errors.extend(parser_diagnostics.deref().clone());

    match fe_analyzer::analyze_ingot(&db, ingot_id) {
        Ok(_) => {}
        Err(diagnostics) => {
            errors.extend(diagnostics.into_iter());
            return Err(CompileError(errors));
        }
    };

    if !errors.is_empty() {
        // There was a non-fatal parser error (eg missing parens in a fn def `fn foo: ...`)
        return Err(CompileError(errors));
    }

    let module_id = ingot_id.main_module(&db).expect("missing main module");

    // build abi
    let json_abis = fe_abi::build(&db, module_id).expect("failed to generate abi");
    let src_ast = format!("{:#?}", &module_id.ast(&db));

    let lowered_ingot_id = fe_lowering::lower_ingot(&db, ingot_id);

    fe_analyzer::analyze_ingot(&db, lowered_ingot_id).expect("failed to analyze lowered AST");

    let lowered_module_id = lowered_ingot_id
        .main_module(&db)
        .expect("missing main module");

    let lowered_ast = format!("{:#?}", &lowered_module_id.ast(&db));

    // compile to yul
    let yul_contracts = fe_yulgen::compile(&db, lowered_module_id);

    // compile to bytecode if required
    #[cfg(feature = "solc-backend")]
    let bytecode_contracts = if _with_bytecode {
        match fe_yulc::compile(yul_contracts.clone(), _optimize) {
            Err(error) => {
                for error in serde_json::from_str::<Value>(&error.0)
                    .expect("unable to deserialize json output")["errors"]
                    .as_array()
                    .expect("errors not an array")
                {
                    eprintln!(
                        "Error: {}",
                        error["formattedMessage"]
                            .as_str()
                            .expect("error value not a string")
                            .replace("\\\n", "\n")
                    )
                }
                panic!("Yul compilation failed with the above errors")
            }
            Ok(contracts) => contracts,
        }
    } else {
        IndexMap::new()
    };

    // combine all of the named contract maps
    let contracts = json_abis
        .keys()
        .map(|name| {
            (
                name.to_owned(),
                CompiledContract {
                    json_abi: json_abis[name].to_owned(),
                    yul: yul_contracts[name].to_owned(),
                    #[cfg(feature = "solc-backend")]
                    bytecode: if _with_bytecode {
                        bytecode_contracts[name].to_owned()
                    } else {
                        "".to_string()
                    },
                },
            )
        })
        .collect::<IndexMap<_, _>>();

    Ok(CompiledModule {
        src_ast,
        lowered_ast,
        contracts,
    })
}
