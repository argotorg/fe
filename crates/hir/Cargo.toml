[package]
name = "fe-hir"
version = "0.20.0-alpha"
authors = ["The Fe Developers <snakecharmers@ethereum.org>"]
edition = "2021"
license = "Apache-2.0"
repository = "https://github.com/ethereum/fe"
description = "Provides HIR definition and lowering for Fe lang"

[dependencies]
tracing = "0.1"
# We may need to fix this to a specific version, 
# but I want to keep up with the latest version until the new Fe implemeentation is merged into the master.
salsa = { git = "https://github.com/salsa-rs/salsa", package = "salsa-2022" }
derive_more = "0.99"
cranelift-entity = "0.91"
num-bigint = "0.4.3"
num-traits = "0.2.15"
camino = "1.1.4"
rustc-hash = "1.1.0"

parser = { path = "../parser2", package = "fe-parser2" }
common = { path = "../common2", package = "fe-common2" }
