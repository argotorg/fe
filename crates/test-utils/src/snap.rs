use std::ops::Range;

pub fn line_col_from_cursor(cursor: parser::TextSize, s: &str) -> (usize, usize) {
    let mut line = 0usize;
    let mut col = 0usize;
    for (i, ch) in s.chars().enumerate() {
        if i == Into::<usize>::into(cursor) {
            return (line, col);
        }
        if ch == '\n' {
            line += 1;
            col = 0;
        } else {
            col += 1;
        }
    }
    (line, col)
}

/// Render a codespan-reporting snippet with primary carets for defs and refs only (no cursor).
pub fn codespan_render_defs_refs(
    file_name: &str,
    content: &str,
    defs: &[(Range<usize>, String)],
    refs: &[(Range<usize>, String)],
) -> String {
    use codespan_reporting::diagnostic::{Diagnostic, Label, Severity};
    use codespan_reporting::term::{emit, Config};
    use termcolor::Buffer;

    let mut out = Buffer::no_color();
    let cfg = Config::default();
    let mut files = codespan_reporting::files::SimpleFiles::new();
    let file_id = files.add(file_name.to_string(), content.to_string());
    let mut labels: Vec<Label<usize>> = Vec::new();
    for (r, msg) in defs.iter() {
        labels.push(Label::primary(file_id, r.clone()).with_message(format!("def: {}", msg)));
    }
    for (r, msg) in refs.iter() {
        labels.push(Label::primary(file_id, r.clone()).with_message(format!("ref: {}", msg)));
    }
    let diag = Diagnostic::new(Severity::Help)
        .with_message("definitions + references")
        .with_labels(labels);
    let _ = emit(&mut out, &cfg, &files, &diag);
    String::from_utf8_lossy(out.as_slice()).into_owned()
}
