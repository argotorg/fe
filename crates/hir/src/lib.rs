use common::{InputDb, Upcast};

pub mod hir_def;
pub mod input;
pub mod lower;
pub mod span;

#[salsa::jar(db = HirDb)]
pub struct Jar(
    // Tracked Hir items.
    hir_def::Mod,
    hir_def::Fn,
    hir_def::ExternFn,
    hir_def::Struct,
    hir_def::Contract,
    hir_def::Enum,
    hir_def::TypeAlias,
    hir_def::Impl,
    hir_def::Trait,
    hir_def::ImplTrait,
    hir_def::Const,
    hir_def::Use,
    // Interned structs.
    hir_def::Body,
    hir_def::IdentId,
    hir_def::IntegerId,
    hir_def::StringId,
    hir_def::PathId,
    hir_def::FnParamListId,
    hir_def::AttrListId,
    hir_def::WhereClauseId,
    hir_def::GenericArgListId,
    hir_def::GenericParamListId,
    hir_def::RecordFieldListId,
    hir_def::EnumVariantListId,
    hir_def::ImplItemListId,
    hir_def::TypeId,
    hir_def::UseTreeId,
    hir_def::ingot_module_tree,
    hir_def::module_item_tree,
);

pub trait HirDb: salsa::DbWithJar<Jar> + InputDb + Upcast<dyn InputDb> {}
impl<DB> HirDb for DB where DB: ?Sized + salsa::DbWithJar<Jar> + InputDb + Upcast<dyn InputDb> {}
