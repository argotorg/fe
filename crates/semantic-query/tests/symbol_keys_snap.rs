use common::InputDb;
use dir_test::{dir_test, Fixture};
use driver::DriverDataBase;
use fe_semantic_query::SemanticQuery;
use hir::{lower::map_file_to_mod, span::LazySpan as _, SpannedHirDb};
use hir_analysis::HirAnalysisDb;
use test_utils::snap::{codespan_render_defs_refs, line_col_from_cursor};
use test_utils::snap_test;
use url::Url;

fn symbol_label<'db>(
    db: &'db dyn SpannedHirDb,
    adb: &'db dyn HirAnalysisDb,
    key: &fe_semantic_query::SymbolKey<'db>,
) -> String {
    use fe_semantic_query::SymbolKey;
    match key {
        SymbolKey::Scope(sc) => sc.pretty_path(db).unwrap_or("<scope>".into()),
        SymbolKey::EnumVariant(v) => v.scope().pretty_path(db).unwrap_or("<variant>".into()),
        SymbolKey::Method(fd) => {
            // Show container scope path + method name
            let name = fd.name(adb).data(db);
            let path = fd.scope(adb).pretty_path(db).unwrap_or_default();
            if path.is_empty() {
                format!("method {}", name)
            } else {
                format!("{}::{}", path, name)
            }
        }
        SymbolKey::FuncParam(item, idx) => {
            let path = hir::hir_def::scope_graph::ScopeId::from_item(*item)
                .pretty_path(db)
                .unwrap_or_default();
            format!("param#{} of {}", idx, path)
        }
        SymbolKey::Local(func, _bkey) => {
            let path = func.scope().pretty_path(db).unwrap_or_default();
            format!("local in {}", path)
        }
    }
}

#[dir_test(dir: "$CARGO_MANIFEST_DIR/test_files", glob: "*.fe")]
fn symbol_keys_snapshot(fx: Fixture<&str>) {
    let mut db = DriverDataBase::default();
    let file = db.workspace().touch(
        &mut db,
        Url::from_file_path(fx.path()).unwrap(),
        Some(fx.content().to_string()),
    );
    let top = map_file_to_mod(&db, file);

    // Modules in this ingot
    let ing = top.ingot(&db);
    let view = ing.files(&db);
    let mut modules: Vec<hir::hir_def::TopLevelMod> = Vec::new();
    for (_u, f) in view.iter() {
        if f.kind(&db) == Some(common::file::IngotFileKind::Source) {
            modules.push(map_file_to_mod(&db, f));
        }
    }
    if modules.is_empty() {
        modules.push(top);
    }

    // Build symbol index across modules
    let map = SemanticQuery::build_symbol_index_for_modules(&db, &modules);

    // Stable ordering of symbol keys via labels
    let mut entries: Vec<(String, fe_semantic_query::SymbolKey)> = map
        .keys()
        .map(|k| (symbol_label(&db, &db, k), k.clone()))
        .collect();
    entries.sort_by(|a, b| a.0.cmp(&b.0));

    let mut out = String::new();
    for (label, key) in entries {
        // Gather def
        let def_opt = SemanticQuery::definition_for_symbol(&db, key)
            .and_then(|(_tm, span)| span.resolve(&db));
        // Gather refs across modules
        let refs = SemanticQuery::references_for_symbol(&db, top, key.clone());
        let mut refs_by_file: std::collections::BTreeMap<
            common::file::File,
            Vec<common::diagnostics::Span>,
        > = Default::default();
        for r in refs {
            if let Some(sp) = r.span.resolve(&db) {
                refs_by_file.entry(sp.file).or_default().push(sp);
            }
        }

        out.push_str(&format!("Symbol: {}\n", label));

        // Group by files that have def or refs
        let mut files: Vec<common::file::File> = refs_by_file.keys().cloned().collect();
        if let Some(d) = def_opt.as_ref() {
            if !files.contains(&d.file) {
                files.push(d.file);
            }
        }
        // Stable order by file URL path
        files.sort_by_key(|f| f.url(&db).map(|u| u.path().to_string()).unwrap_or_default());

        for f in files {
            let content = f.text(&db);
            let name = f
                .url(&db)
                .and_then(|u| {
                    u.path_segments()
                        .and_then(|mut s| s.next_back())
                        .map(|s| s.to_string())
                })
                .unwrap_or_else(|| "<file>".into());
            let mut defs_same: Vec<(std::ops::Range<usize>, String)> = Vec::new();
            let mut refs_same: Vec<(std::ops::Range<usize>, String)> = Vec::new();

            if let Some(def) = def_opt.as_ref().filter(|d| d.file == f) {
                let s: usize = Into::<usize>::into(def.range.start());
                let e: usize = Into::<usize>::into(def.range.end());
                let (l0, c0) = line_col_from_cursor(def.range.start(), &content);
                let (l, c) = (l0 + 1, c0 + 1);
                // total refs count across all files
                let total_refs = refs_by_file.values().map(|v| v.len()).sum::<usize>();
                defs_same.push((
                    s..e,
                    format!("defined here @ {}:{} ({} refs)", l, c, total_refs),
                ));
            }

            if let Some(v) = refs_by_file.get(&f) {
                let mut spans = v.clone();
                spans.sort_by_key(|sp| (sp.range.start(), sp.range.end()));
                for sp in spans {
                    let s: usize = Into::<usize>::into(sp.range.start());
                    let e: usize = Into::<usize>::into(sp.range.end());
                    let (l0, c0) = line_col_from_cursor(sp.range.start(), &content);
                    let (l, c) = (l0 + 1, c0 + 1);
                    refs_same.push((s..e, format!("{}:{}", l, c)));
                }
            }

            // Render codespan for this file for this symbol
            let block = codespan_render_defs_refs(&name, &content, &defs_same, &refs_same);
            out.push_str(&block);
            out.push('\n');
        }

        out.push('\n');
    }

    let orig = std::path::Path::new(fx.path());
    let stem = orig
        .file_stem()
        .and_then(|s| s.to_str())
        .unwrap_or("snapshot");
    let combined_name = format!("{}.snap", stem);
    let combined_path = orig.with_file_name(combined_name);
    snap_test!(out, combined_path.to_str().unwrap());
}
