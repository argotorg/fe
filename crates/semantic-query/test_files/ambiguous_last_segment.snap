---
source: crates/semantic-query/tests/symbol_keys_snap.rs
expression: out
input_file: test_files/ambiguous_last_segment.snap
---
Symbol: ambiguous_last_segment::m
help: definitions + references
  ┌─ ambiguous_last_segment.fe:1:5
  │
1 │ mod m {
  │     ^
  │     │
  │     def: defined here @ 1:5 (2 refs)
  │     ref: 1:5
  ·
6 │ use m::ambiguous
  │     ^ ref: 6:5



Symbol: ambiguous_last_segment::m::ambiguous
help: definitions + references
  ┌─ ambiguous_last_segment.fe:2:12
  │
2 │     pub fn ambiguous() {}
  │            ^^^^^^^^^
  │            │
  │            def: defined here @ 2:12 (2 refs)
  │            ref: 2:12
  ·
6 │ use m::ambiguous
  │        ^^^^^^^^^ ref: 6:8
